File: Advanced/PageAndRDL/RtfControl/VB.NET/Rtf/Control/RichTextBoxFixed.vb
Imports System.Text
Imports System.Windows.Forms

Imports GrapeCity.ActiveReports.Samples.Rtf.Native

Namespace Control
	Public Class RichTextBoxFixed
		Inherits RichTextBox

		Public Overrides Property Text As String
			Get
				Dim textLength As Integer = GetTextLength(GTL_USECRLF)
				Dim getText = New GETTEXTEX()
				getText.cb = textLength + 2
				getText.flags = GT_USECRLF
				getText.codepage = 1200
				Dim sb = New StringBuilder(CType(getText.cb / 2, Integer))
				UnsafeNativeMethods.SendMessage(Handle, EM_GETTEXTEX, getText, sb)
				Return sb.ToString()
			End Get

			Set(ByVal value As String)
				MyBase.Text = value
			End Set
		End Property

		Public Overrides ReadOnly Property TextLength As Integer
			Get
				Return GetTextLength(GTL_USECRLF) / 2
			End Get
		End Property

		Public Overrides Property SelectionLength As Integer
			Get
				Dim charRange = New CHARRANGE()
				UnsafeNativeMethods.SendMessage(Handle, EM_EXGETSEL, 0, charRange)
				Return charRange.cpMax - charRange.cpMin
			End Get

			Set(ByVal value As Integer)
				MyBase.SelectionLength = value
			End Set
		End Property

		Private Function GetTextLength(ByVal gtl As Integer) As Integer
			Dim getLength = New GETTEXTLENGTHEX()
			getLength.flags = gtl
			getLength.codepage = 1200
			Return UnsafeNativeMethods.SendMessage(Handle, EM_GETTEXTLENGTHEX, getLength, 0)
		End Function
	End Class
End NameSpace

File: Advanced/PageAndRDL/RtfControl/C#/Rtf/Control/RichTextBoxFixed.cs
using System.Text;
using System.Windows.Forms;

using GrapeCity.ActiveReports.Samples.Rtf.Native;

namespace GrapeCity.ActiveReports.Samples.Rtf.Control
{
	public class RichTextBoxFixed : RichTextBox
	{
		public override string Text
		{
			get
			{
				int textLength = GetTextLength(NativeMethods.GTL_USECRLF);

				var getText = new NativeMethods.GETTEXTEX();
				getText.cb = textLength + 2;

				getText.flags = NativeMethods.GT_USECRLF;
				getText.codepage = 1200;

				var sb = new StringBuilder(getText.cb / 2);

				UnsafeNativeMethods.SendMessage(Handle, NativeMethods.EM_GETTEXTEX, ref getText, sb);

				return sb.ToString();
			}
			set { base.Text = value; }
		}

		public override int TextLength
		{
			get { return GetTextLength(NativeMethods.GTL_USECRLF) / 2; }
		}

		public override int SelectionLength
		{
			get
			{
				var charRange = new NativeMethods.CHARRANGE();
				UnsafeNativeMethods.SendMessage(Handle, NativeMethods.EM_EXGETSEL, 0, ref charRange);
				return charRange.cpMax - charRange.cpMin;
			}
			set { base.SelectionLength = value; }
		}

		private int GetTextLength(int gtl)
		{
			var getLength = new NativeMethods.GETTEXTLENGTHEX();
			getLength.flags = gtl;
			getLength.codepage = 1200;
			return UnsafeNativeMethods.SendMessage(Handle, NativeMethods.EM_GETTEXTLENGTHEX, ref getLength, 0);
		}
	}
}
